File: python/JAPL/lexer.py
from .meta.tokenobject import Token
from .meta.tokentype import TokenType
from .meta.exceptions import ParseError
from typing import List


class Lexer(object):
    """
       A simple tokenizer for the JAPL programming
       language, scans a input source file and
       produces a list of tokens. Some errors
       are caught here as well.
    """

    TOKENS = {"(": TokenType.LP, ")": TokenType.RP,
              "{": TokenType.LB, "}": TokenType.RB,
              ".": TokenType.DOT, ",": TokenType.COMMA,
              "-": TokenType.MINUS, "+": TokenType.PLUS,
              ";": TokenType.SEMICOLON, "*": TokenType.STAR,
              ">": TokenType.GT, "<": TokenType.LT,
              "=": TokenType.EQ, "!": TokenType.NEG,
              "/": TokenType.SLASH, "%": TokenType.MOD}

    RESERVED = {"or": TokenType.OR, "and": TokenType.AND,
                "class": TokenType.CLASS, "fun": TokenType.FUN,
                "if": TokenType.IF, "else": TokenType.ELSE,
                "for": TokenType.FOR, "while": TokenType.WHILE,
                "var": TokenType.VAR, "nil": TokenType.NIL,
                "true": TokenType.TRUE, "false": TokenType.FALSE,
                "return": TokenType.RETURN,
                "this": TokenType.THIS, "super": TokenType.SUPER,
                "del": TokenType.DEL, "break": TokenType.BREAK}

    def __init__(self, source: str):
        """Object constructor"""

        self.source = source
        self.tokens: List[Token] = []
        self.line: int = 1   # Points to the line being lexed
        self.start: int = 0  # The position of the first character of the current lexeme
        self.current: int = 0  # The position of the current character being lexed

    def step(self) -> str:
        """
        'Steps' one character in the source code and returns it
        """

        if self.done():
            return ""
        self.current += 1
        return self.source[self.current - 1]

    def peek(self) -> str:
        """
        Returns the current character without consuming it
        or an empty string if all text has been consumed
        """

        if self.done():
            return ""
        return self.source[self.current]

    def peek_next(self) -> str:
        """
        Returns the next character after self.current
        or an empty string if the input has been consumed
        """

        if self.current + 1 >= len(self.source):
            return ""
        return self.source[self.current + 1]

    def string(self, delimiter: str):
        """Parses a string literal"""

        while self.peek() != delimiter and not self.done():
            if self.peek() == "\n":
                self.line += 1
            self.step()
        if self.done():
            raise ParseError(f"unterminated string literal at line {self.line}")
        self.step()   # Consume the '"'
        value = self.source[self.start + 1:self.current - 1]  # Get the actual string
        self.tokens.append(self.create_token(TokenType.STR, value))

    def number(self):
        """Parses a number literal"""

        convert = int
        while self.peek().isdigit():
            self.step()
        if self.peek() == ".":
            self.step()  # Consume the "."
            while self.peek().isdigit():
                self.step()
            convert = float
        self.tokens.append(self.create_token(TokenType.NUM,
                                             convert(self.source[self.start:self.current])))

    def identifier(self):
        """Parses identifiers and reserved keywords"""

        while self.peek().isalnum() or self.is_identifier(self.peek()):
            self.step()
        kind = TokenType.ID
        value = self.source[self.start:self.current]
        if self.RESERVED.get(value, None):
            kind = self.RESERVED[value]
        self.tokens.append(self.create_token(kind))

    def comment(self):
        """Handles multi-line comments"""

        closed = False
        while not self.done():
            end = self.peek() + self.peek_next()
            if end == "/*":   # Nested comments
                self.step()
                self.step()
                self.comment()
            elif end == "*/":
                closed = True
                self.step()   # Consume the two ends
                self.step()
                break
            self.step()
        if self.done() and not closed:
            raise ParseError(f"Unexpected EOF at line {self.line}")

    def match(self, char: str) -> bool:
        """
        Returns True if the current character in self.source matches
        the given character
        """

        if self.done():
            return False
        elif self.source[self.current] != char:
            return False
        self.current += 1
        return True

    def done(self) -> bool:
        """
        Helper method that's used by the lexer
        to know if all source has been consumed
        """

        return self.current >= len(self.source)

    def create_token(self, kind: TokenType, literal: object = None) -> Token:
        """
        Creates and returns a token object
        """

        return Token(kind, self.source[self.start:self.current], literal, self.line)

    def is_identifier(self, char: str):
        """Returns if a character can be an identifier"""

        if char.isalpha() or char in ("_", ):  # More coming soon
            return True

    def scan_token(self):
        """
        Scans for a single token and adds it to
        self.tokens
        """

        char = self.step()
        if char in (" ", "\t", "\r"):  # Useless characters
            return
        elif char == "\n":   # New line
            self.line += 1
        elif char in ("'", '"'):   # A string literal
            self.string(char)
        elif char.isdigit():
            self.number()
        elif self.is_identifier(char):  # Identifier or reserved keyword
            self.identifier()
        elif char in self.TOKENS:
            if char == "/" and self.match("/"):
                while self.peek() != "\n" and not self.done():
                    self.step()   # Who cares about comments?
            elif char == "/" and self.match("*"):
                self.comment()
            elif char == "=" and self.match("="):
                self.tokens.append(self.create_token(TokenType.DEQ))
            elif char == ">" and self.match("="):
                self.tokens.append(self.create_token(TokenType.GE))
            elif char == "<" and self.match("="):
                self.tokens.append(self.create_token(TokenType.LE))
            elif char == "!" and self.match("="):
                self.tokens.append(self.create_token(TokenType.NE))
            elif char == "*" and self.match("*"):
                self.tokens.append(self.create_token(TokenType.POW))
            else:
                self.tokens.append(self.create_token(self.TOKENS[char]))
        else:
            raise ParseError(f"unexpected character '{char}' at line {self.line}")

    def lex(self) -> List[Token]:
        """
        Performs lexical analysis on self.source
        and returns a list of tokens
        """

        while not self.done():
            self.start = self.current
            self.scan_token()
        self.tokens.append(Token(TokenType.EOF, "", None, self.line))
        return self.tokens

File: lexer.nim
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at

#  http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.


## A simple tokenizer implementation with one character of lookahead.
## This module has been designed to be easily extendible in its functionality
## given that JAPL is in a state of high activity and many features are
## being added along the way. To add support for a new keyword, just create
## an appropriate TokenType entry in the enum in the file at meta/tokentype.nim
## and then add it to the constant RESERVED table. A similar approach applies for
## other tokens, but multi-character ones require more tweaking.
## Since this lexer scans the given source string character by character, unicode
## identifiers are not supported

import system
import strutils
import strformat
import tables
import meta/tokentype
import meta/tokenobject
import meta/valueobject

# Table of all tokens except reserved keywords
const TOKENS = to_table({
              '(': TokenType.LP, ')': TokenType.RP,
              '{': TokenType.LB, '}': TokenType.RB,
              '.': TokenType.DOT, ',': TokenType.COMMA,
              '-': TokenType.MINUS, '+': TokenType.PLUS,
              ';': TokenType.SEMICOLON, '*': TokenType.STAR,
              '>': TokenType.GT, '<': TokenType.LT,
              '=': TokenType.EQ, '!': TokenType.NEG,
              '/': TokenType.SLASH, '%': TokenType.MOD,
              '[': TokenType.LS, ']': TokenType.RS,
              ':': TokenType.COLON, '^': TokenType.CARET,
              '&': TokenType.BAND, '|': TokenType.BOR,
              '~': TokenType.TILDE})

# Constant table storing all the reserved keywords for JAPL
const RESERVED = to_table({
                "or": TokenType.OR, "and": TokenType.AND,
                "class": TokenType.CLASS, "fun": TokenType.FUN,
                "if": TokenType.IF, "else": TokenType.ELSE,
                "for": TokenType.FOR, "while": TokenType.WHILE,
                "var": TokenType.VAR, "nil": TokenType.NIL,
                "true": TokenType.TRUE, "false": TokenType.FALSE,
                "return": TokenType.RETURN,
                "this": TokenType.THIS, "super": TokenType.SUPER,
                "del": TokenType.DEL, "break": TokenType.BREAK,
                "continue": TokenType.CONTINUE, "inf": TokenType.INF,
                "nan": TokenType.NAN})
type
    Lexer* = ref object
        source*: string
        tokens*: seq[Token]
        line*: int
        start*: int
        current*: int
        errored*: bool
        file*: string


func initLexer*(source: string, file: string): Lexer =
    ## Initializes the lexer
    result = Lexer(source: source, tokens: @[], line: 1, start: 0, current: 0, errored: false, file: file)


proc done(self: Lexer): bool =
    ## Returns true if we reached EOF
    result = self.current >= self.source.len


proc step(self: var Lexer): char =
    ## Steps one character forward in the
    ## source file. A null terminator is returned
    ## if the lexer is at EOF
    if self.done():
        return '\0'
    self.current = self.current + 1
    result = self.source[self.current - 1]


proc peek(self: Lexer): char =
    ## Returns the current character in the
    ## source file without consuming it.
    ## A null terminator is returned
    ## if the lexer is at EOF
    if self.done():
        result = '\0'
    else:
        result = self.source[self.current]


proc match(self: var Lexer, what: char): bool =
    ## Returns true if the next character matches
    ## the given character, and consumes it.
    ## Otherwise, false is returned
    if self.done():
        return false
    elif self.peek() != what:
        return false
    self.current += 1
    return true


proc peekNext(self: Lexer): char =
    ## Returns the next character
    ## in the source file without
    ## consuming it.
    ## A null terminator is returned
    ## if the lexer is at EOF
    if self.current + 1 >= self.source.len:
        result = '\0'
    else:
        result = self.source[self.current + 1]


proc createToken(self: var Lexer, tokenType: TokenType, literal: Value): Token =
    ## Creates a token object for later use in the parser
    result = Token(kind: tokenType,
                   lexeme: self.source[self.start..<self.current],
                   literal: literal,
                   line: self.line
                   )


proc parseString(self: var Lexer, delimiter: char) =
    ## Parses string literals
    while self.peek() != delimiter and not self.done():
        if self.peek() == '\n':
            self.line = self.line + 1
        discard self.step()
    if self.done():
        stderr.write(&"A fatal error occurred while parsing '{self.file}', line {self.line} at '{self.peek()}' -> Unterminated string literal\n")
        self.errored = true
    discard self.step()
    let value = self.source[self.start..<self.current].asStr() # Get the value between quotes
    let token = self.createToken(STR, value)
    self.tokens.add(token)


proc parseNumber(self: var Lexer) =
    ## Parses numeric literals
    # TODO: Move the integer conversion to the
    # compiler for finer error reporting and
    # handling
    while isDigit(self.peek()):
        discard self.step()
    try:
        if self.peek() == '.':
            discard self.step()
            while self.peek().isDigit():
                discard self.step()
            var value = parseFloat(self.source[self.start..<self.current]).asFloat()
            self.tokens.add(self.createToken(TokenType.NUMBER, value))
        else:
            var value = parseInt(self.source[self.start..<self.current]).asInt()
            self.tokens.add(self.createToken(TokenType.NUMBER, value))
    except ValueError:
        stderr.write(&"A fatal error occurred while parsing '{self.file}', line {self.line} at '{self.peek()}' -> integer is too big to convert to int64\n")
        self.errored = true


proc parseIdentifier(self: var Lexer) =
    ## Parses identifiers, note that
    ## multi-character tokens such as
    ## UTF runes are not supported
    while self.peek().isAlphaNumeric():
        discard self.step()
    var text: string = self.source[self.start..<self.current]
    var keyword = text in RESERVED
    if keyword:
        self.tokens.add(self.createToken(RESERVED[text], text.asStr()))
    else:
        self.tokens.add(self.createToken(ID, text.asStr()))


proc parseComment(self: var Lexer) =
    ## Parses multi-line comments. They start
    ## with /* and end with */, and can be nested.
    ## A missing comment terminator will raise an
    ## error
    var closed = false
    while not self.done():
        var finish = self.peek() & self.peekNext()
        if finish == "/*":   # Nested comments
            discard self.step()
            discard self.step()
            self.parseComment()   # Recursively parse any other enclosing comments
        elif finish == "*/":
            closed = true
            discard self.step()   # Consume the two ends
            discard self.step()
            break
        discard self.step()
    if self.done() and not closed:
        self.errored = true
        stderr.write(&"A fatal error occurred while parsing '{self.file}', line {self.line} at '{self.peek()}' -> Unexpected EOF\n")


proc scanToken(self: var Lexer) =
    ## Scans a single token. This method is
    ## called iteratively until the source
    ## file reaches EOF
    var single = self.step()
    if single in [' ', '\t', '\r']:  # We skip whitespaces, tabs and other useless characters
        return
    elif single == '\n':
        self.line += 1
    elif single in ['"', '\'']:
        self.parseString(single)
    elif single.isDigit():
        self.parseNumber()
    elif single.isAlphaNumeric() or single == '_':
        self.parseIdentifier()
    elif single in TOKENS:
        if single == '/' and self.match('/'):
            while self.peek() != '\n' and not self.done():
                discard self.step()
        elif single == '/' and self.match('*'):
            self.parseComment()
        elif single == '=' and self.match('='):
            self.tokens.add(self.createToken(DEQ, "==".asStr()))
        elif single == '>' and self.match('='):
            self.tokens.add(self.createToken(GE, ">=".asStr()))
        elif single == '>' and self.match('>'):
            self.tokens.add(self.createToken(SHR, ">>".asStr()))
        elif single == '<' and self.match('='):
            self.tokens.add(self.createToken(LE, "<=".asStr()))
        elif single == '<' and self.match('<'):
            self.tokens.add(self.createToken(SHL, ">>".asStr()))
        elif single == '!' and self.match('='):
            self.tokens.add(self.createToken(NE, "!=".asStr()))
        elif single == '*' and self.match('*'):
            self.tokens.add(self.createToken(POW, "**".asStr()))
        else:
            self.tokens.add(self.createToken(TOKENS[single], asStr(&"{single}")))
    else:
        self.errored = true
        stderr.write(&"A fatal error occurred while parsing '{self.file}', line {self.line} at '{self.peek()}' -> Unexpected token '{single}'\n")


proc lex*(self: var Lexer): seq[Token] =
    ## Lexes a source file, converting a stream
    ## of characters into a series of tokens
    while not self.done():
        self.start = self.current
        self.scanToken()
    self.tokens.add(Token(kind: EOF, lexeme: "EOF", literal: Value(kind: ValueType.Nil), line: self.line))
    return self.tokens

